src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs

```diff
@@ -55,16 +56,19 @@ public ValueTask<int> RequestCreditAsync(int amount)
                     return new ValueTask<int>(granted);
                 }
 
-                var tcs = new TaskCompletionSource<int>(TaskContinuationOptions.RunContinuationsAsynchronously);
+                // Uses RunContinuationsAsynchronously internally.
+                var tcs = new TaskCompletionSourceWithCancellation<int>();
 
                 if (_waiters == null)
                 {
                     _waiters = new Queue<Waiter>();
                 }
 
-                _waiters.Enqueue(new Waiter { Amount = amount, TaskCompletionSource = tcs });
+                Waiter waiter = new Waiter { Amount = amount, TaskCompletionSource = tcs };
 
-                return new ValueTask<int>(tcs.Task);
+                _waiters.Enqueue(waiter);
+
+                return new ValueTask<int>(tcs.WaitWithCancellationAsync(cancellationToken));
```

Comment:
I'd suggest making this:
```C#
return new ValueTask<int>(cancellationToken.CanBeCanceled ?
    tcs.WaitWithCancellationAsync(cancellationToken) :
    tcs.Task;
```
Otherwise, if the cancellation token might have been `CancellationToken.None`/`default(CancellationToken)`, we'll end up paying additional unnecessary costs (e.g. an extra async method state machine).

Reply:
I'll fix this for the next commit. Thanks.